extruct/rdfa.py

```diff
@@ -28,6 +29,100 @@
+            'cc'    : 'http://creativecommons.org/ns#',
+            'grddl'   : 'http://www.w3.org/2003/g/data-view#',
+            'rif'   : 'http://www.w3.org/2007/rif#',
+            'sioc'    : 'http://rdfs.org/sioc/ns#',
+            'skos'    : 'http://www.w3.org/2004/02/skos/core#',
+            'xml'   : 'http://www.w3.org/XML/1998/namespace',
+            'rdfs'    : 'http://www.w3.org/2000/01/rdf-schema#',
+            'rev'   : 'http://purl.org/stuff/rev#',
+            'rdfa'    : 'http://www.w3.org/ns/rdfa#',
+            'dc'    : 'http://purl.org/dc/terms/',
+            'foaf'    : 'http://xmlns.com/foaf/0.1/',
+            'void'    : 'http://rdfs.org/ns/void#',
+            'ical'    : 'http://www.w3.org/2002/12/cal/icaltzd#',
+            'vcard'   : 'http://www.w3.org/2006/vcard/ns#',
+            'wdrs'    : 'http://www.w3.org/2007/05/powder-s#',
+            'og'    : 'http://ogp.me/ns#',
+            'wdr'   : 'http://www.w3.org/2007/05/powder#',
+            'rdf'   : 'http://www.w3.org/1999/02/22-rdf-syntax-ns#',
+            'xhv'   : 'http://www.w3.org/1999/xhtml/vocab#',
+            'xsd'   : 'http://www.w3.org/2001/XMLSchema#',
+            'v'     : 'http://rdf.data-vocabulary.org/#',
+            'skosxl'  : 'http://www.w3.org/2008/05/skos-xl#',
+            'schema'  : 'http://schema.org/',
+        }
+
+        # if bad property
+        if ':' not in prop:
+            return prop
+        
+        # if property has no prefix
+        if 'http://' in prop:
+            return prop
+
+        prefix = prop.split(':')[0]
+        
+        match = None
+        if head_element.get('prefix'):
+            match = re.search(prefix + ': [^\s]+', head_element.get('prefix'))
+
+        # if namespace taken from prefix attribute in head tag
+        if match:
+            ns = match.group().split(': ')[1]
+            return ns + prop.split(':')[1]
+
+        # if namespace taken from xmlns attribute in html tag
+        if ('xmlns:' + prefix) in html_element.keys():
+            return html_element.get('xmlns:' + prefix) + prop.split(':')[1]
+
+        # if namespace present in inital context
+        if prefix in context:
+            return context[prefix] + prop.split(':')[1]
+
+        return prop
+    
+    # sorts the rdfa tags in jsonld string
+    def sort(self, unordered, ordered):
+        idx_for_value = dict((value, idx) for idx, value in enumerate(ordered))
+        unordered.sort(key=lambda props: idx_for_value.get(props.get('@value'), len(ordered)))
+    
+    
+    # fixes order of rdfa tags in jsonld string
+    # by comparing with order in document object
+    def fixOrder(self, jsonld_string, document):
+        html_element = document.xpath('/html')[0]
+        head_element = document.xpath('//head')[0]
+                
+        for meta_tag in head_element.xpath("meta[@property]"):
+            meta_tag.attrib['property'] = self.replaceNS(meta_tag.attrib['property'], 
```

Comment:
I see now a problem here. This is actually modifying the `document` itself, and can have then side effects in other parts of the code. I think it can be avoided by just building a `defaultdict` with the properties and its content, and use it below in the the code. Could you please change it?

Reply:
I see the problem now. I am not sure how to use `defaultdict` though. Will making a deep copy of the `document` object be sufficient?

Reply:
Making a copy of the document is very inefficient. You don't need a `defaultdict`, you can use a simple `dict` for it but `defaultdict` makes your life easier: https://www.accelebrate.com/blog/using-defaultdict-python#:~:text=Dictionaries%20are%20a%20convenient%20way,retrieval%20by%20name%20(key).&text=A%20defaultdict%20works%20exactly%20like,will%20never%20raise%20a%20KeyError.